src/samples/audio/detect_test.c

```diff
@@ -363,85 +377,123 @@ static void test_keyword_free(struct comp_dev *dev)
+static int test_keyword_set_config(struct comp_dev *dev, char *data,
+				   uint32_t data_size)
 {
-	struct comp_data *cd = comp_get_drvdata(dev);
-	struct comp_buffer *sourceb;
-	struct comp_buffer __sparse_cache *source_c;
-	unsigned int channels, rate;
-	enum sof_ipc_frame frame_fmt;
-	int err;
+	struct sof_detect_test_config *cfg;
+	size_t cfg_size;
 
-	/* Detector is used only in KPB topology. It always requires channels
-	 * parameter set to 1.
-	 */
-	params->channels = 1;
+	/* Copy new config */
+	cfg = (struct sof_detect_test_config *)data;
+	cfg_size = data_size;
 
-	err = test_keyword_verify_params(dev, params);
-	if (err < 0) {
-		comp_err(dev, "test_keyword_params(): pcm params verification failed.");
+	comp_info(dev, "test_keyword_set_config(): config size = %u",
+		  cfg_size);
+
+	if (cfg_size != sizeof(struct sof_detect_test_config)) {
+		comp_err(dev, "test_keyword_set_config(): invalid config size");
 		return -EINVAL;
 	}
 
-	cd->sample_valid_bytes = params->sample_valid_bytes;
+	return test_keyword_apply_config(dev, cfg);
+}
 
-	/* keyword components will only ever have 1 source */
-	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
-				  sink_list);
-	source_c = buffer_acquire(sourceb);
-	channels = source_c->stream.channels;
-	frame_fmt = source_c->stream.frame_fmt;
-	rate = source_c->stream.rate;
-	buffer_release(source_c);
+static int test_keyword_get_config(struct comp_dev *dev, char *data,
+				   uint32_t *data_size)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	size_t cfg_size;
+	int ret = 0;
 
-	if (channels != 1) {
-		comp_err(dev, "test_keyword_params(): only single-channel supported");
-		return -EINVAL;
-	}
+	comp_info(dev, "test_keyword_get_config()");
 
-	if (!detector_is_sample_width_supported(frame_fmt)) {
-		comp_err(dev, "test_keyword_params(): only 16-bit format supported");
+	cfg_size = sizeof(struct sof_detect_test_config);
+
+	if (cfg_size > *data_size) {
+		comp_info(dev, "test_keyword_get_config(): wrong config size: %d",
+			  *data_size);
 		return -EINVAL;
 	}
 
-	/* calculate the length of the preamble */
-	if (cd->config.preamble_time) {
-		cd->keyphrase_samples = cd->config.preamble_time *
-					(rate / 1000);
-	} else {
-		cd->keyphrase_samples = KEYPHRASE_DEFAULT_PREAMBLE_LENGTH;
-	}
+	*data_size = cfg_size;
 
-	err = test_keyword_get_threshold(dev, params->sample_valid_bytes * 8);
-	if (err < 0) {
-		comp_err(dev, "test_keyword_params(): unsupported sample width %u",
-			 params->sample_valid_bytes * 8);
-		return err;
+	/* Copy back to user space */
+	ret = memcpy_s(data, cfg_size, &cd->config, cfg_size);
+	assert(!ret);
+
+	return ret;
+}
+
+static int test_keyword_set_large_config(struct comp_dev *dev,
+					 uint32_t param_id,
+					 bool first_block,
+					 bool last_block,
+					 uint32_t data_offset,
+					 char *data)
+{
+	comp_dbg(dev, "test_keyword_set_large_config()");
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	assert(cd);
```

Comment:
We should only be using assert for unrecoverable errors (as it kills all audio). Fwiw, this check is not needed here as new() will allocate and assign here. i.e. it will always be valid.